### PR TITLE
executor: buffer snapshotChunk to avoid too many object memory allocation

### PR DESCRIPTION
**Pull request by XuHuaiyu**
PTAL @jackysp @coocood

### CODE REVIEW

**Comment by winoros** [comment]
Post some benchmark result?

**Comment by jackysp** [comment]
qps: 14k -> 22k

**Comment by XuHuaiyu** [comment]
PTAL @coocood @zz-jason @jackysp

File: executor/union_scan.go [hunk 1]
```diff
@@ -106,11 +106,23 @@ type UnionScanExec struct {
...
+	snapshotChunkBuffer    *chunk.Chunk
+	it4SnapshotChunkBuffer *chunk.Iterator4Chunk
+}
+
+// Open implements the Executor Open interface.
+func (us *UnionScanExec) Open(ctx context.Context) error {
+	if err := us.baseExecutor.Open(ctx); err != nil {
+		return errors.Trace(err)
+	}
+	us.snapshotChunkBuffer = chunk.NewChunk(us.retTypes())
```
Comment (zz-jason) [inline executor/union_scan.go]: why not just call `us.newChunk()` ?

File: executor/union_scan.go [hunk 2]
```diff
@@ -106,11 +106,23 @@ type UnionScanExec struct {
...
+	it4SnapshotChunkBuffer *chunk.Iterator4Chunk
+}
+
+// Open implements the Executor Open interface.
+func (us *UnionScanExec) Open(ctx context.Context) error {
+	if err := us.baseExecutor.Open(ctx); err != nil {
+		return errors.Trace(err)
+	}
+	us.snapshotChunkBuffer = us.newChunk()
+	us.it4SnapshotChunkBuffer = chunk.NewIterator4Chunk(us.snapshotChunkBuffer)
```
Comment (winkyao) [inline executor/union_scan.go]: Why iterator need buffer?

File: session/bench_test.go [hunk 3]
```diff
@@ -341,3 +347,20 @@ func BenchmarkJoinLimit(b *testing.B) {
 		readResult(ctx, rs[0], 1)
 	}
 }
+
+func BenchmarkUnionScan(b *testing.B) {
```
Comment (winkyao) [inline session/bench_test.go]: what about the BenchmarkUnionScan result?

File: executor/union_scan.go [hunk 4]
```diff
@@ -182,13 +194,12 @@ func (us *UnionScanExec) getSnapshotRow(ctx context.Context) (types.DatumRow, er
 	us.cursor4SnapshotRows = 0
 	us.snapshotRows = us.snapshotRows[:0]
 	for len(us.snapshotRows) == 0 {
-		chk := chunk.NewChunkWithCapacity(us.retTypes(), us.maxChunkSize)
-		err = us.children[0].Next(ctx, chk)
-		if err != nil || chk.NumRows() == 0 {
+		us.snapshotChunkBuffer.Reset()
```
Comment (coocood) [inline executor/union_scan.go]: This line can be removed

**Pull request merged**